### PR TITLE
cppcheck performance fixes in interfaces/

### PR DESCRIPTION
**Pull request by tobbi**

### CODE REVIEW

**Comment by MartijnKaijser** [comment]
@Montellese @topfs2 ping

File: xbmc/interfaces/legacy/Monitor.h [hunk 1]
```diff
@@ -161,7 +161,7 @@ namespace XBMCAddon
        *\n
        * Will be called when XBMC receives or sends a notification\n
        */
-      virtual void    onNotification(const String sender, const String method, const String data) { XBMC_TRACE; }
+      virtual void    onNotification(const String& sender, const String& method, const String& data) { XBMC_TRACE; }
```
Comment (ace20022) [inline xbmc/interfaces/legacy/Monitor.h]: I guess these changes yield a compile error. You should changes the implementations at line 42ff accordingly, imo.
Haven't you compile checked your changes?

File: xbmc/interfaces/python/PyContext.cpp [hunk 2]
```diff
@@ -93,7 +93,7 @@ namespace XBMCAddon
 
       if (cur->gilReleasedDepth == 0) // true if we are at the outermost
       {
-        PyThreadState* _save;
+        PyThreadState* _save = NULL;
```
Comment (ace20022) [inline xbmc/interfaces/python/PyContext.cpp]: not a performance thing (again)

File: xbmc/interfaces/legacy/Monitor.h (hunk 1 above)
Comment (tobbi) [inline xbmc/interfaces/legacy/Monitor.h]: Actually, I didn't. But I will do this for future pulls.

**Comment by tobbi** [comment]
Please check again, I think I fixed everything now. Compiling Kodi went rather smoothly.

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by MartijnKaijser** [comment]
Last time this one failed but reports are gone. Please check
 jenkins build this please

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by tobbi** [comment]
All I can say is that it builds fine for me (on Mac OS Yosemite). Can you please persist the Jenkins report somehow so that I can take a look even when I am absent for longer?

**Comment by Memphiz** [comment]
not possible - you'd need to copy the output locally if you wanna review it later - atm we have only 3 logs for each build because of diskspace constraints... (will change back to 10 soon - but thats also gone fast...)

**Comment by tobbi** [comment]
Okay, then please rebuild one more time so that I can see what's wrong and then save it somehow.

**Comment by mkortstiege** [comment]
jenkins build this please

**Comment by kib** [comment]
@tobbi grab the logs while they are hot: http://jenkins.kodi.tv/view/Automation/job/BuildMulti-PR/2136/
All builds failing on AddonModuleXbmc.cpp

File: xbmc/interfaces/json-rpc/JSONServiceDescription.cpp [hunk 3]
```diff
@@ -1154,9 +1154,9 @@ void JSONSchemaTypeDefinition::Set(const JSONSchemaTypeDefinitionPtr typeDefinit
   referencedTypeSet = true;
 }
 
-JSONSchemaTypeDefinition::CJsonSchemaPropertiesMap::CJsonSchemaPropertiesMap()
+JSONSchemaTypeDefinition::CJsonSchemaPropertiesMap::CJsonSchemaPropertiesMap():
+   m_propertiesmap(std::map<std::string, JSONSchemaTypeDefinitionPtr>())
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/JSONServiceDescription.cpp]: Does this really need to be so explicit? Wouldn't `m_propertiesmap()` be enough?
Personally I'd also prefer it if either the colon would be on the new line followed by a space or at least add a space between the `)` and the colon for easier readability.

**Comment by tobbi** [comment]
Seems like swig generates code that is incompatible with the changes above. I could either find out how to circumvent the issues or leave out the Monitor changes entirely.

**Comment by ace20022** [comment]
Maybe something like: http://stackoverflow.com/questions/9122511/using-swig-with-methods-that-take-stdstring-as-a-parameter .

**Comment by MartijnKaijser** [comment]
Maybe split up the PR so the obvious can go in

**Comment by tobbi** [comment]
I have done that now. Will check the rest soon.

**Comment by Paxxi** [comment]
Not sure about this. Generally I prefer this method of initiliazation but doing function calls and logic there can probably lead to all sorts of weirdness. Ordering issues seems the most likely. As for perf gains, are they even measurable?

File: xbmc/interfaces/legacy/WindowInterceptor.h [hunk 4]
```diff
@@ -124,9 +124,9 @@ namespace XBMCAddon
 
     public:
       Interceptor(const char* specializedName,
-                  Window* _window, int windowid) : P(windowid, "")
-      { 
-        ((classname = "Interceptor<") += specializedName) += ">";
+                  Window* _window, int windowid) : P(windowid, ""),
+        classname("Interceptor<" + (std::string)specializedName + ">")
```
Comment (Paxxi) [inline xbmc/interfaces/legacy/WindowInterceptor.h]: Casting char* to std::string looks weird, I would go for std::string(specializedName)

Comment (tobbi) [inline xbmc/interfaces/legacy/WindowInterceptor.h]: done.

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by tobbi** [comment]
Build failures seem unrelated as far as I can see.

**Comment by anaconda** [comment]
> Build failures seem unrelated as far as I can see.
This builds fine on the Pi (as confirmed by Jenkins). I’ve been testing this for a few hours and everything seems okay.

**Comment by MartijnKaijser** [comment]
jenkins build this please to be sure

**Comment by Paxxi** [comment]
In general this looks good but I object to the changes in Control.cpp. Initialization logic(function calls, control flow) should be in the contructor body and not in the initializer list. I'd like to see these changes reverted before this PR is merged.

**Comment by mkortstiege** [comment]
What @xhaggi wrote. @tobbi please remove the Control.cpp changes and rebase so we can trigger a merge build.

**Comment by tobbi** [comment]
Undid Control.cpp changes.

**Comment by mkortstiege** [comment]
Thanks. jenkins build and merge

**Pull request merged**